### ISSUE TITLE
S3 Bucket Naming Limitations

### ISSUE THREAD

**Issue by nanodust** [s3fs-fuse/s3fs-fuse#434]
According to Amazon: 
The rules for bucket names in the US East (N. Virginia) region allow bucket names to be as long as 255 characters, and bucket names can contain any combination of uppercase letters, lowercase letters, numbers, periods (.), hyphens , and underscores (_).

Yet, with S3FS, I cannot use any uppercase, periods, hyphens, or underscores in S3FS - if I do, I get an "s3fs:  bucket name contains an illegal character"

this is true regardless of whether or not I use -ouse_path_request_style 

In summary - s3fs works fine on buckets with no period, hyphen, underscore.

If any of those exist, I cannot find a way to allow s3fs to work. 

I'm writing this issue in case it helps others...

meanwhile if there is a way to actually use (at least hyphens!) please let us know. 

Thank you !

**Comment by andrewgaul** [comment]
Which version of s3fs do you use?  I successfully created and mounted such a bucket with 1.80 via:

```
s3fs Gaul-_bucket $MOUNTPOINT -o createbucket -o use_path_request_style
```

**Comment by nanodust** [comment]
thanks Andrew. I'm fresh / latest : 

Amazon Simple Storage Service File System V1.80(commit:36917f7) with OpenSSL

it only works when I don't have hyphens or underscores in my bucket name. 

As I can't re-name the buckets (i need underscores & hyphens), I just moved over to aws cli sync, and gave up on fuse (which seemed the more elegant solution).

**Comment by r14c** [comment]
I am seeing this issue on 1.89 when i try to mount a bucket using fstab.

**Comment by michaelwclark** [comment]
I'm having this issue as well. Manually mounting or fstab

```Amazon Simple Storage Service File System V1.84(commit:unknown) with GnuTLS(gcrypt)
Copyright (C) 2010 Randy Rizun <rrizun@gmail.com>
License GPL2: GNU GPL version 2 <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.```

**Comment by DavidGarciaCat** [comment]
Me too... this seems like a recurrent issue that should not be closed...

```
$ sudo vi /etc/fstab

s3fs#dgc-test /opt/test/dgc fuse.s3fs _netdev,allow_other,passwd_file=/etc/passwd-s3fs,default_acl=public-read,uid=1000,gid=1000 0 0

$ sudo mount -a

s3fs: BUCKET s3fs#dgc-test -- bucket name contains an illegal character.

$ s3fs --version

Amazon Simple Storage Service File System V1.86 (commit:unknown) with GnuTLS(gcrypt)
Copyright (C) 2010 Randy Rizun <rrizun@gmail.com>
License GPL2: GNU GPL version 2 <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by 0xRook1e** [comment]
same error in /etc/fstab

**Comment by gaul** [comment]
Please test with the latest version 1.91.  We do not support older versions.

**Comment by tve** [comment]
Same here. I can mount a bucket that has a hyphen using the `s3fs` command but not via fstab. Version 1.92:

```
/s3> s3fs --version
Amazon Simple Storage Service File System V1.92 (commit:unknown) with OpenSSL
Copyright (C) 2010 Randy Rizun <rrizun@gmail.com>
```

E.g. this works:
```
/s3> sudo s3fs js-board /s3/js-board/ -f -d -o passwd_file=/etc/passwd-s3fs            
```
And this doesn't:
```
/s3> sudo mount js-board
s3fs: BUCKET /s3/js-board -- bucket name contains an illegal character.
```
This is the fstab entry (wrapped for readability):
```
js-board /s3/js-board fuse.s3fs noatime,_netdev,nonempty,passwd_file=/etc/passwd-s3fs,\
default_acl=public-read,use_cache=/play/tmp/s3-cache,del_cache,\
url=https://s3.us-west-2.amazonaws.com,gid=1000,uid=1000,allow_other 0 0
```

**Comment by wrenchpilot** [comment]
I was having this same issue, the documentation on what to put into /etc/fstab is not great. Here's what worked for me. **Don't put anything before the bucket name.**

```bash
BUCKETNAME /MOUNTPOINT fuse.s3fs _netdev,allow_other,passwd_file=/root/.passwd-s3fs 0 0
```

**Comment by dMailonG** [comment]
@gaul could you please reopen this issue?

**Comment by dMailonG** [comment]
Just tested it, and it seems related to the bucket and the mount directory having the same name.
This is further reinforced by the error message exhibiting the full path to the mount directory as the bucket name.

Bucket name: my-bucket
Directory: /media/mailon/my-bucket
Result: Fail (s3fs: BUCKET /media/mailon/my-bucket -- bucket name contains an illegal character.)

Bucket name: mybucket
Directory: /media/mailon/my-bucket
Result: Success

Bucket name: my-bucket
Directory: /media/mailon/mybucket
Result: Success

As per **CheckForbiddenBucketParams()** on **s3fs_cred.cpp**, dashes are not illegal characters, but slashes are:
```
    // check bucket name for illegal characters
    size_t found = bucket_name.find_first_of("/:\\;!@#$%^&*?|+=");
    if(found != std::string::npos){
        S3FS_PRN_EXIT("BUCKET %s -- bucket name contains an illegal character: '%c' at position %zu", bucket_name.c_str(), bucket_name[found], found);
        return false;
    }
```
I'd say the issue arises around **my_fuse_opt_proc()** on **s3fs.cpp**, but I'm unfamiliar with C++, so it's hard for me to tell.

**Comment by DavidGarciaCat** [comment]
Hello @gaul 

I was wondering if you or one of your colleagues maintaining `s3fs-fuse` have had a chance to review @dMailonG's previous post (written about 2.5 months ago), as it seems to identify an existing issue in one of the `cpp` files.

Thanks,

**Comment by ggtakec** [comment]
@DavidGarciaCat @dMailonG 
I tried this with the current master branch code, but I can't reproduce it.

I tried the path style case and the other case, and both were mounted without any problems.
The fstab entry I tried is the following:
```
(1) my-bucket /mnt/my-bucket fuse.s3fs _netdev,allow_other,passwd_file=/etc/passwd-s3fs 0 0
(2) my-bucket /mnt/my-bucket fuse.s3fs _netdev,allow_other,passwd_file=/etc/passwd-s3fs,url=https://s3-ap-northeast-1.amazonaws.com 0 0
```
If you are still having trouble with this problem, specify the `dbglevel` and `curldbg` options as follows to have s3fs output debug logs.
These debug logs are output to files under `/var/log`, etc. (If you don't know the log file, please refer to the `logfile` option.)
```
my-bucket /mnt/my-bucket fuse.s3fs _netdev,allow_other,passwd_file=/etc/passwd-s3fs,dbglevel=dbg,curldbg 0 0
```
Then, if the problem persists, please post a new issue with that information(debug logs, etc.).

Thanks in advance for your assistance.

**Comment by kallisti5** [comment]
ps: any further lurkers... this one has to do with your current working directory. If you move out of the path where the fuse mount happens, the error goes away

Conditions:
* /etc/fstab    (example mounting to: /s3/blah-thing)
* your pwd is /s3
* you ```mount -a```

In the conditions above, you get "bucket name contains an illegal character"
